packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -423,14 +423,26 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
                     if (this.hostPolicy.concurrentSnapshotFetch && !this.hostPolicy.summarizerClient) {
                         const snapshotP = this.fetchSnapshot(hostSnapshotOptions);
 
-                        const promiseRaceWinner = await promiseRaceWithWinner([cachedSnapshotP, snapshotP]);
+                        // Ensure that failures on both paths are ignored initially.
+                        // I.e. if cache fails for some reason, we will proceed with network result.
+                        // And vice versa - if (for example) client is offline and network request fails first, we
+                        // do want to attempt to succeed with cached data!
+                        const promiseRaceWinner = await promiseRaceWithWinner([
+                            cachedSnapshotP.catch(() => undefined),
+                            snapshotP.catch(() => undefined),
+                        ]);
                         cachedSnapshot = promiseRaceWinner.value;
 
-                        if (cachedSnapshot === undefined) {
-                            cachedSnapshot = await snapshotP;
-                        }
-                        else {
-                            snapshotP.catch(() => {});
+                        if (cachedSnapshot === undefined)
+                        {
+                            // if network failed -> wait for cache ( then return network failure)
+                            // If cache returned empty or failed -> wait for network (success of failure)
+                            if (promiseRaceWinner.index === 1) {
+                                cachedSnapshot = await cachedSnapshotP;
+                            }
+                            if (cachedSnapshot === undefined) {
+                                cachedSnapshot = await snapshotP;
+                            }
                         }
 
```

Comment:
Might be clearer to introduce a new 'raceToFirstSuccess' helper.  Something like the (untested) code below...

```ts
function firstSuccess(promises: Promise<unknown>[], errorCallback: (error, index) => boolean) {
    return new Promise((resolve, reject) => {
        for (let index = 0; index < promises.length; index++) {
            promises[index].then(
                // First successful result resolves the race (if not already rejected)
                (value) => resolve({ index, value })
            ).catch((error: unknown) => {
                // Caller will receive callback for all errors, even after success.
                if (errorCallback(error, index)) { reject(error) };
            });
        }
    });
}
```

Reply:
I'm not sure how I'd use it. I.e. callback can't do much synchronously, as we need to wait for the other promise (in case it succeeds).